Sync Page title with most recently captured version

maxtedford opened the pull request:
Resolves #59 

This work adds a `title` attribute to the `versions` table, and supplies an `after_add` callback to `Page` model, which sets the `Page`'s `title` attribute to match that of the `Version` with the most recent `capture_time`.

I'm not quite familiar enough with the application at this point to know intuitively whether the `title` attribute on a `Version` should include a validation for presence as a mandatory field, but my best judgement led me to forego that addition, and simpy match it to the `Page`'s implementation of the same field.

Please let me know if I've missed anything or if there's anything else I can do to help on this ticket! 😄

Mr0grog commented:
This sounds great, thanks so much @maxtedford! I am super busy this afternoon, but will try and look at this with a closer eye tonight or tomorrow at the latest.

Mr0grog commented on app/models/page.rb:
```diff
@@ -38,4 +42,11 @@ def url_must_have_domain
+  def sync_titles(version)
+    if version.title.present?
+      most_recent_capture_time = versions.limit(1).order(capture_time: :desc).pluck(:capture_time).first
+      self.title = version.title if most_recent_capture_time.nil? || most_recent_capture_time <= version.capture_time
```
You need to also call `save!` here to save this change to the DB. You could also call `update` instead of `title=` (`update` immediately saves):

```rb
update(title: version.title) if most_recent_capture_time.nil? || most_recent_capture_time <= version.capture_time
```

See below comment in the tests about making sure the changes are saved, too.

Mr0grog commented on app/models/page.rb:
```diff
@@ -38,4 +42,11 @@ def url_must_have_domain
+
+  def sync_titles(version)
+    if version.title.present?
+      most_recent_capture_time = versions.limit(1).order(capture_time: :desc).pluck(:capture_time).first
```
It might be easier/clearer to just call `latest` here:

```rb
most_recent_capture_time = latest.capture_time
```

Mr0grog commented on db/migrate/20170909151007_add_title_to_versions.rb:
```diff
@@ -0,0 +1,5 @@
+class AddTitleToVersions < ActiveRecord::Migration[5.1]
+  def change
+    add_column :versions, :title, :string
```
I think we should also make sure we fill in the `title` attribute on all existing versions with the existing title from their page. You could do that here or make a second migration for it.

maxtedford replied:
Makes sense to me. I just went ahead and added that to this existing migration.

Mr0grog commented on test/models/page_test.rb:
```diff
@@ -13,4 +13,24 @@ class PageTest < ActiveSupport::TestCase
+    assert_equal('Page One', page.title)
+
+    page.versions.create(title: 'Newest Version', capture_time: '2017-03-05T00:00:00Z')
+    assert_equal('Newest Version', page.title, 'The page title should always sync against the title of the version with the most recent capture time')
```
Alongside this assertion, you also need to make sure the changes were persisted to the DB (see note in the model file). You should get a fresh copy of the page from the DB to make assertions on:

```rb
page.versions.create(title: 'Newest Version', capture_time: '2017-03-05T00:00:00Z')

# Get a fresh page instance
page = Page.find(page.uuid) # OR: page.reload
assert_equal('Newest Version', page.title, '...')
# etc.
```

You could also check that the page wasn’t left with unsaved changes:

```rb
assert_not(page.changed?, 'The page was left with unsaved changes')
```

Mr0grog commented on test/models/page_test.rb:
```diff
@@ -13,4 +13,24 @@ class PageTest < ActiveSupport::TestCase
+    assert_equal('Page One', page.title)
+
+    page.versions.create(capture_time: '2017-03-05T00:00:00Z')
+    assert_equal('Page One', page.title, 'The page title should not sync with the incoming version if it has a nil title')
```
Same issue about page saving as above.

Mr0grog commented:
Also, pro tip: if you put “fixes #59” (or “closes” or “resolves”) instead of “delivers,” GitHub will automatically close the issue when this is merged! It’s super convenient. More info: https://help.github.com/articles/closing-issues-using-keywords/

maxtedford commented:
Hey @Mr0grog, thanks for the thorough review! I'm totally embarrassed that I didn't actually save the record as a part of the `after_add` callback. I think in my head I was treating it as a `before_action`, and thus assumed an automatic save as a part of the process. But you're absolutely right, so I updated it accordingly, including the tests.

I didn't even notice the `latest` relation you had on the `Page` model. That's super helpful! As I was making updates to the `import_versions_job` to no longer create new pages with a `title` attribute, I considered adding a method alias to simply pull the `title` from the `latest` record (something like the below)

```ruby
alias_method :original_title, :title

def title
end
```

, but decided instead to simply update the test to `assert_nil pages[0].title` and `assert_equal [

Mr0grog commented on db/migrate/20170909151007_add_title_to_versions.rb:
```diff
@@ -0,0 +1,8 @@
+  def change
+    add_column :versions, :title, :string
+    Version.includes(:page).find_each do |v|
+      puts "Updated Version record with UUID #{v.uuid} to have title '#{v.title}'" if v.update(title: v.page.title)
```
I think this is going to be verbose enough to be a problem (we have ~335,000 versions in production right now). This much logging will actually slow the migration way down and probably not be helpful. Maybe better to count the number of changes and just output a summary, e.g. “Updated title on X version records for Y pages.”

Also, this should probably only run when migrating up, not down, so maybe move this part into a `reversible` block with only an `up` section?

Finally, Rails has some nice logging helpers for migrations. You could do something like:

```rb
say_with_time("Copying titles from pages to versions") do
  Version.includes(:page).find_each do |v|
    # Update version and count it
  end
  say("X versions", :subitem)
  say("Y pages", :subitem)
end
```

…and you get nice output like:

```
== 20170909151007 AddTitleToVersions: migrating ===============================
-- add_column(:versions, :title, :string)
   -> 0.0013s
-- Copying titles from pages to versions
   -> 1019 versions
   -> 146 pages
   -> 2.3985s
== 20170909151007 AddTitleToVersions: migrated (2.3999s) ======================
```

Mr0grog replied:
Oh, also: this might be faster if you roll through the list of pages and call `update_all` on their versions; that way there’s only one SQL query per page, rather than one for every version, e.g:

```rb
Page.each do |page|
  page.versions.update_all(title: page.title)
end
```

`update_all` also returns the number of records that were affected, which is helpful in doing the summary counting I mentioned above.

maxtedford replied:
Very cool! I've never made use of these Rails helpers before, so TIL. I'll change that up.

Mr0grog commented on test/controllers/api/v0/imports_controller_test.rb:
```diff
@@ -64,9 +64,10 @@ def teardown
     assert_equal import_data[0][:site_agency], pages[0].agency
     assert_equal import_data[0][:site_name], pages[0].site
+    assert_equal import_data[0][:page_title], pages[0].latest.title
+    assert_nil pages[0].title
```
Shouldn’t this *not* be `nil`? i.e. the earlier assertion (`assert_equal import_data[0][:page_title], pages[0].title`) should still be here, shouldn’t it?

Mr0grog replied:
We should probably also change the import data here to `title` instead of `page_title`, or have tests that use both (since we’re transitioning from one to the other).

maxtedford replied:
Yes, it makes sense that this test should still pass as is. I dug a little more into this, and it seems that [this line](https://github.com/edgi-govdata-archiving/web-monitoring-db/blob/master/app/jobs/import_versions_job.rb#L69) wasn't actually triggering the `after_add` callback on the `Page` model. Interestingly enough, if I move the callback into the `Version` model in the form of an `after_save`, it does get triggered, so I'll push up that change as well.

maxtedford replied:
I'll also revert the test back. I should have trusted the test as it was originally written.

Mr0grog replied:
Oh, I think the more immediate problem was actually https://github.com/edgi-govdata-archiving/web-monitoring-db/blob/master/app/jobs/import_versions_job.rb#L52, which should have been `page.versions.push(version)`

maxtedford replied:
Ah, yep I think you're right!

Mr0grog commented on test/models/page_test.rb:
```diff
@@ -13,4 +13,32 @@ class PageTest < ActiveSupport::TestCase
+    page.versions.create(capture_time: '2017-03-05T00:00:00Z')
+    page.reload
+    assert_equal('Page One', page.title, 'The page title should not sync with the incoming version if it has a nil title')
+    refute(page.changed?, 'The page was left with unsaved changes')
```
…I *think* this is pointless after you’ve called `reload` (it trashes all state on the model object). Not totally sure on that, though.

maxtedford replied:
I also looked into this, and you're right. The `reload` method _does_ in fact trash any existing state on the model object. Also interestingly, however, I discovered (after throwing a couple `binding.pry` calls between a few of these assertions) that we don't actually need to call `reload` to get the new state on these objects.

```ruby
binding.pry
page.versions.create(title: 'Newest Version', capture_time: '2017-03-05T00:00:00Z')
binding.pry
```

`page.changed?` returned `false` from inside both of those `binding.pry`s despite the `page` record having actually had it's `title` attribute updated. So, it would seem we don't need to worry about calling `reload` to confirm these changes.

Mr0grog replied:
I think this is because you changed the callback to use `update`, which saves immediately, so you *should* get `page.changed? == false`. I definitely tested this with an assertion before you made the latest changes and it did in fact fail. I think it would be good to have that assertion to guard against any regression there.

maxtedford replied:
Makes sense to me. I can add it back in as a regression safeguard 👍

Mr0grog commented:
> I considered adding a method alias on the Page model to simply pull the title from the latest record in the event of a nil title (something like the below)… but decided instead to simply stick with what you recommended

I think that was a good choice. The virtual attribute would have been convenient, but potentially a big problem for large queries that return many pages (you don’t want those to wind up issuing an additional SQL query for each record). It probably also would have required changes everywhere we JSON serialize pages.

> `assert_nil pages[0].title`

Noted this inline, too, but I’m not sure this makes sense. Shouldn’t the pages still be acquiring the title from its versions? So this shouldn’t be `nil` when all is said and done.

maxtedford commented:
Hey @Mr0grog, I just pushed up a few more small changes based on your latest feedback. Please let me know if there's anything else I can do to help out!

Mr0grog commented on app/models/version.rb:
```diff
@@ -8,11 +8,22 @@ class Version < ApplicationRecord
     class_name: 'Change',
     foreign_key: 'uuid_to'
 
+  after_save :sync_page_title
```
I’m a little worried that this will be a lot higher overhead than `after_add` on page association, but it’s probably fine. We could do `after_create`, though.

maxtedford replied:
Yeah, it probably will seeing as versions are saved in more contexts than just when they're added to a page. I'm obviously happy to go with whichever you think is best, either `after_save` or `after_create`. For the time being, I'll change it to `after_create`, but let me know if you prefer I leave it be.

maxtedford commented:
Okay, I've restored the `refute(page.changed?)` to those page model tests as a safeguard, changed the `after_save` on the `Version` model to an `after_create`, and made that offering to the 🤖 👮 gods. Let me know if you notice anything else! 🤞

Mr0grog commented:
Beautiful, thank you so much for working all the way through this! 🎉 💃
